### ISSUE TITLE
paramfetch should allow operators to download only the parameters which they'll use

### ISSUE THREAD

**Issue by laser** [filecoin-project/rust-fil-proofs#629]
### Description

Developers today use 256MiB and 1024KiB sectors when sealing. The devnet/user cluster uses 256MiB sectors when sealing. Soon, we will publish Groth parameters for 1GiB sectors. 

Developers will likely not use this sector size during their normal development flow. They should not have to download Groth parameters for 1GiB sectors if they do not plan to seal 1GiB sectors.

Mining rig operators will use only one sector size. They should not have to download Groth parameters for sector sizes they'll never use when sealing.

### Acceptance criteria

- prevent go-filecoin operators and developers from needing to download all published Groth parameters
- ensure that all verifying keys are downloaded regardless of which Groth parameters are downloaded

### Risks + pitfalls

- will likely need to modify `parampublish` such that `parameters.json` includes sector size

--------

# Proposal: Supporting 1GiB Sectors

For additional background, see [this](https://github.com/filecoin-project/rust-fil-proofs/issues/629) and [this](https://github.com/filecoin-project/rust-fil-proofs/issues/625).

## Problem

Developers use 256MiB and 1KiB sectors today. Cryptolab envisions publishing 1GiB sectors sometime in the near future, and I think that we should start running large sectors in the devnet/user cluster within the next month or two so that we can see where our system fails.

We don't want to require go-filecoin developers or mining operators to download Groth parameters for sectors which they never intend to seal. Our existing `paramfetch` program allows an operator to either download all parameters or download parameters on a per filename basis. The file names don't communicate the _meaning_ of the Groth parameter-file (e.g. the filename `v10-zigzag-proof-of-replication-6648cd509799c8b76f944bbc971692c69d1480fa60c569bfc1284f0bd8ead988` doesn't communicate the size of the sector), so an operator doesn't have the information required to opt out of downloading Groth parameters for sectors which they'll never seal.

### Additional Metadata in `parameters.json`

The existing `parameters.json` file should be modified to include additional parameter and verifying key-related metadata. `parampublish` will need to be modified in order to achieve this. We will augment each JSON object in `parameters.json` to include the number of bytes in the associated (sealed) sector:

```json
{
  "v10-vdf-post-c8ddc8e74f9c2ed6a4bc83df171a9e9e84e242cf6e48e20f71585106fa8a3ce6.vk": {
    "sector_size": 266338304,
    "cid": "QmTKL1MiWSWFUMEECZPaBX6ddoZy9HCAA4mdaW65BKDoEp",
    "digest": "2684b87a68d3a02118584e524a3f073a"
  },
  "v10-zigzag-proof-of-replication-6648cd509bc971692c69d1480fa60c569bfc1284f0bd8ead988": {
    "sector_size": 1024,
    "cid": "QmeAwbo2CM4t58kV7byr1X2beA7pzbGvStCVSDSG4ydu2E",
    "digest": "a30003a7b2b2512cd5e66cfefdd9fe7b"
  },
  "v10-vdf-post-912e5a392c09cbfa0bf369ac77f939a45e0f2aec968cf3846cbe356a19b76685": {
    "sector_size": 1024,
    "cid": "QmP1PkagfiC9cUEh3r7NEWaJfnBGEnqfABxXxh3PSoiMtX",
    "digest": "ff7fa52e9bf2a3b5261f6d5740156f79"
  }
  // ...
}
```

### New `paramfetch` flag

The `paramfetch` program should be modified to accept a comma-separated list of sector sizes:

```shell
./paramfetch --sector-sizes=1024
```

The `--sector-sizes` flag must be incompatible with the `--all` flag. If the `--sector-sizes` flag is provided and its value is a well-formed list of sector sizes, `paramfetch` should download all verifying keys (distinguished by their `.vk` suffix) and all non-verifying-key-files with a `sector_size` property whose value exists in the `--sector-sizes` list. It should do this without prompting the user; behavior should be similar in this regard to the `--all` flag.

### Improved `paramfetch` Prompt

The `paramfetch` program, when run without `--all` or `--sector-sizes` flags, should include in its prompt the size of the sector:

```shell
[y/n] (sector_size: 266338304B) v10-zigzag-proof-of-replication-6648cd509799c8b76f944b92c69d1480fa60c569bfc1284f0bd8ead988.vk
```

**Comment by porcuquine** [comment]
Quick note that everyone *will* need to download all verifying keys. However, these will not be large.

**Comment by laser** [comment]
@porcuquine Agreed. That's part of the acceptance criteria, above.

**Comment by porcuquine** [comment]
So it is. Sorry for the oversight. Carry on.

**Comment by laser** [comment]
Lightweight proposal can be found [here](https://docs.google.com/document/d/16-asJsvqL94WY5F_lzjBPsNVWjJoUcmY-ISMb2A9xbo/edit).

**Comment by vmx** [comment]
@laser Shouldn't the lightweight proposal be public, so everyone can work on it?

**Comment by laser** [comment]
@vmx - Ah, yes, but of course I will! Link [here](https://docs.google.com/document/d/16-asJsvqL94WY5F_lzjBPsNVWjJoUcmY-ISMb2A9xbo/edit?usp=sharing).

**Comment by dignifiedquire** [comment]
@laser for simplicity, can we move the proposal into this issue?

**Comment by anorth** [comment]
Just wanted to note that downloading 1.2GB parameter files is a big pain and I would really like this improvement. The large files take me about 20 minutes each. Due to the call from go-filecoin's `make deps` it tends to happen at a highly inopportune time like when updating a PR for landing, or sitting down to some focussed dev work. I never use large sectors in development.

(Other suggestions for decoupling from go-filecoin build are also welcome).

**Comment by porcuquine** [comment]
I agree we should do this, but out of curiosity, are you forced to download parameters often? I think they only changed once in the the month of May. If you need to download more often than when they change it makes me think something strange is going on.

**Comment by laser** [comment]
If the default directory (a subdirectory of /tmp, using OSX) is used, parameters will be wiped each time your default directory is cleared out. On my machine, this happens each time I restart.

**Comment by porcuquine** [comment]
This might be the root UX problem (which doesn't discount the other issue, of course). I think intended use is for Filecoin node operators (including developers) to customize the value of the FILECOIN_PARAMETER_CACHE environment variable so the parameters will be cached in a location that has the characteristics they desire.

**Comment by anorth** [comment]
I reboot as infrequently as possible. I am here again 14 days later waiting for downloads which are only 1/4 the size expected with 1GB sectors.

**Comment by porcuquine** [comment]
A few thoughts:

@laser For generality, we should also add proof-size metadata, since it's possible to have parameters/verification-keys for the same sector size but which vary by proof size.

@anorth Can you clarify what the issue related to the 14 day interval is? Are you saying that's when you had to reboot and were forced to download again? I ask because I really think having to 're-download' parameters is a separate issue which also needs to be addressed independently.

Regardless of which order the issues are addressed, arranging for an appropriate parameter directory will definitely be necessary — especially as the number of parameters increases. No one is going to want to download the same parameters unnecessarily, so this is independent of parameter selection. It's also the change which will address the currently reported problem. To be clear, I'm not arguing against the parent issue — just clarifying that if anyone is repeatedly downloading groth parameters, that is a problem in its own right.

**Comment by laser** [comment]
> @laser For generality, we should also add proof-size metadata, since it's possible to have parameters/verification-keys for the same sector size but which vary by proof size.

@porcuquine - I have created an issue, #705 (_Groth parameter cache must include proof partitions_), which blocks this issue.

**Comment by laser** [comment]
> I reboot as infrequently as possible. I am here again 14 days later waiting for downloads which are only 1/4 the size expected with 1GB sectors.

@anorth - Setting the `FILECOIN_PARAMETER_CACHE` to path which is not removed when you reboot will prevent you from having to re-download a parameter/VK pair which you have previously downloaded. Do you think that the parameter cache-hydrating/downloading tools should use a directory which is not `/tmp/filecoin-proof-parameters` by default? Your location makes the existing UX particularly painful. How should we improve the UX for this tool?

Also: Cryptolab periodically publishes new parameters. Our tooling forces developers to download all parameters when they are published (regardless of which parameters that developer plans to use). Looking at the [history of parameters.json](https://github.com/filecoin-project/rust-fil-proofs/commits/master/parameters.json), it looks like we publish once or twice a month. When this issue is complete, a developer will be able to opt in to downloading parameters - but the cache-clearing (when the developer wipes their `/tmp` directory) remains.

**Comment by anorth** [comment]
My computer informs me that it has been up for 48 days, so the two most recent times I've complained in this issue, 17 June and 3 June, were probably not caused by that. I noted "14 days" just as a counterpoint to your claim of "once a month". Maybe that was the cause of three downloads ago?

The UX of losing giant and necessary files on reboot does seem sufficiently bad that it's worth adding up-front friction to avoid it. I suggest either defaulting to a directory that will not have that behaviour (may require system-dependent switching) or, simpler, not having a default at all and requiring the user to select one explicitly. This would have the advantage of communicating to the user that these files exist and are important and large. It could mesh well with the opt-in requirement for fetching them in the first place.

**Comment by anorth** [comment]
Garrrr, I was forced to restart and here I am again. 

As far as I can tell, `FILECOIN_PARAMETER_CACHE` is completely undocumented so I don't expect anyone uses it. I don't know if this issue is the right place, but again this likelihood of losing the giant files is super frustrating. I'm setting up that env variable now, but strongly maintain that the user should be forced to chose a permanent location up front.

**Comment by anorth** [comment]
I was going to update the wiki to set `FILECOIN_PARAMETER_CACHE` when fetching parameters, but will hold off in hope of a better flow overall.

I fear that if the var is set for downloading but not also set at every invocation of `go-filecoin daemon` then the daemon will regenerate the parameters from scratch, which is even worse (#649)!

**Comment by schomatis** [comment]
(Sorry to divert the focus of this issue but could we standardize the env.var through our new settings manager in a future release, renaming it as `FIL_PROOFS_PARAMETER_CACHE`? Happy to submit a PR about it. I fear that at some point in the future someone will confuse both prefixes and the setting might get lost in the noise.)

**Comment by laser** [comment]
@anorth @porcuquine - I just spent a few hours digging in to the Groth parameter cache populating and reading code. Here's a brain dump, along with a proposed solution. 

Please do read the "Challenges" section, as it provides some insight into why setting a non-`/tmp/...` default from within `go-filecoin` source code is not going to work.

-------

# Groth Parameter/Verifying Keys

Storage miners need Groth parameters in order to generate PoSt and seal proofs. A storage miner needs one Groth parameter file per combination of (sector size, circuit, proof partitions) for which they wish to generate a proof. As of the time I write this (July 4, 2019), a Groth parameter file is approximately 1.6GiB in size.

A verifying key is needed for verifying the proofs created by storage miners. There exists a single verifying key per Groth parameter. Verifying keys are much smaller than Groth parameters - approximately 1KiB. Anyone wishing to verify a seal or PoSt proof will need the corresponding verifying key.

## Generating and Publishing Groth Parameters + Verifying Keys

Cryptolab typically generates and publishes new parameters and verifying keys when they produce a new proof construction (e.g. Beacon PoSt), when they modify an existing construction (e.g. they change VDF PoSt challenge count), or when they add support for a new sector size.

### Generating: `paramcache`

To generate new parameters and keys, a member of the Cryptolab team runs the `paramcache` program, defined in the `rust-fil-proofs` Git repo. 

The program's source code contains an enumeration of parameters and verifying keys that it knows how to generate, identified by a checksum of the parameter metadata (challenge count, sector size, etc.). The program's source code is part of the same Git repository as the proofs-generating code; `paramcache` built from `rust-fil-proofs` Git SHA `XYZ` will always produces the Groth parameters and verifying keys required by `libsector_builder_ffi.a` built from `rust-fil-proofs` Git SHA `XYZ`. 

The `paramcache` program, by default, checks the `/tmp/filecoin-proof-parameters` directory for the existence of the Groth parameters and keys that it knows how to generate. The `FILECOIN_PARAMETER_CACHE` environment variable can be used to override this default. If any Groth parameter or verifying key  (identified by a checksum of parameter metadata) is missing from the directory, it is generated immediately. On my laptop, this process can take approximately 30 minutes per Groth parameter.

### Publishing: `parampublish`

The `parampublish` program is used to add Groth parameters and verifying keys to a locally-running IPFS daemon. Additionally, it updates the `parameters.json` manifest (found in the root of the `rust-fil-proofs` Git repository and also distributed with the prebuilt `go-filecoin` binary) to include the enumeration of parameters and keys which Cryptolab expects node operators to use.

The `parampublish` program, by default, enumerates the files in the `/tmp/filecoin-proof-parameters` directory and prompts the user to add each file to a locally running IPFS daemon. The `FILECOIN_PARAMETER_CACHE` environment variable can be used to override this default. 
 If `parampublish` exits successfully, the `parameters.json` file will have been updated to include the CID, filename, BLAKE2b checksum, and sector size for each Groth parameter and verifying key added to the locally-running IPFS daemon. A Cryptolab team member will then use the IPFS Pinbot to pin each CID on the IPFS Gateway. Once each CID in the `parameters.json` file has been successfully pinned to the Gateway, publishing is complete.

## Non-Cryptolab Interactions with Groth Parameters and Verifying Keys

### Downloading: `paramfetch`

Developers (those who build `go-filecoin` from source) and end users (those who consume the pre-built `go-filecoin` releases), before starting the go-filecoin daemon, use the `paramfetch` binary to download Groth parameters and verifying keys from the IPFS Gateway. In the case of developers, this happens as a side effect of running `deps` (which in turn calls `paramfetch`). End users will call `paramfetch` directly.

By default, parameters are keys are downloaded to `/tmp/filecoin-proof-parameters`. This default can be overridden by setting the `FILECOIN_PARAMETER_CACHE` environment variable before `paramfetch` runs.

Note: End users will have acquired the `paramfetch` binary and `parameters.json` manifest from the `go-filecoin` release tarball. Developers will have acquired `paramfetch` and `parameters.json` from a `rust-fil-proofs` GitHub release (pinned by submodule SHA).

### Consuming: `libsector_builder_ffi.a`

When a Filecoin developer or end user wishes to seal a sector, the underlying `rust-fil-proofs` code will attempt to load the appropriate Groth parameter from disk. When a user wishes to verify a proof, it will attempt to load the corresponding verifying key from disk. 

This software obeys the same rules as `paramfetch`, `parampublish`, and `paramcache`: If `FILECOIN_PARAMETER_CACHE` specifies a directory, it will look for Groth parameters and keys in that directory. If the environment variable is not set, `/tmp/filecoin-proof-parameters` is used.

## Challenges

Largely, the challenge of choosing a good parameter cache directory default which is not a subdirectory of `/tmp` arises from the fact that end users will not use `deps`, but rather will invoke `paramfetch` directly.

The decision to write parameters to a subdirectory of `/tmp` has solved some problems:

1. We could be reasonably sure that this directory **existed** and was **writeable** in all environments in which `go-filecoin` and `paramfetch` are run. (Note: This is not true of `GOPATH`, which may not be set on systems running the prebuilt `go-filecoin` binary).
1. Using an absolute path as a default meant that we could easily coordinate between disparate binaries (`paramcache`, `go-filecoin`) run from different working directories.
1. Periodic wipes of `/tmp` by the OS meant that we could punt on cache eviction.  ...but also introduced others:  1. Periodic wipes of `/tmp` result in redundant downloads (or generation) of Groth parameters and keys, frustrating developers and wasting time.
1. Some operating systems hold the contents of `/tmp` in RAM, so large Groth parameters lock up lots of memory.

## Potential Solutions

### Environment Variable

To ensure that `paramfetch` and `libsector_builder_ffi.a` (by way of `go-filecoin`) agree on where to write and read Groth parameters and keys, a developer or end user can set the `FILECOIN_PARAMETER_CACHE` environment variable:

To avoid having to set the environment variable as part of each shell command (Bash):

```shell
echo 'export FILECOIN_PARAMETER_CACHE="/Users/flarp/.parameter-cache"' >> ~/.bash_profile
```

### Require Cache Path Argument

As an alternative to reading from an environment variable, we could modify `paramcache`, `parampublish`, and `paramfetch` to require a `--cache-path` flag:

```shell
$ paramfetch --all
Error: Please specify a --cache-path

$ paramfetch --all --cache-path=/Users/flarp/.parameter-cache
# ... downloading
```

The Rust code which produces `libsector_builder_ffi.a` and `libfilecoin_proofs_ffi.a` would need to be modified to include an FFI-exposed `set_cache_path` function which would set the location from which parameters and keys would be read. If that function were not called, the system would behave as it does today.

The `go-filecoin` code would need to be modified to call this FFI-exposed function using a value loaded from its configuration. That configuration value would also need to be exposed such that consumers of the prebuilt `go-filecoin` binary could obtain it for their own call to `paramfetch`.

The `rust-fil-proofs` Rust code would need to be modified such that the downstream crate (`rust-fil-sector-builder`) could manipulate the cache path-holding singleton.

### Configuration File

Alternatively, we could modify `paramcache`, `parampublish`, and `paramfetch` programs to read from a configuration file in either their current working directory.

The `rust-fil-proofs` code (consumed by `libsector_builder_ffi.a`) would also be modified to look for the same configuration file in its current working directory, which it would use to configure the parameter cache directory.

If no configuration file were found in the current working directory of whichever process consumes `libsector_builder_ffi.a`, the `libsector_builder_ffi.a` library would behave as it does today. Same thing for `paramfetch` and `parampublish` and `paramcache`.

## Recommendation

Of the aforementioned solutions, I recommend the approach which uses an environment variable set once in the user's `.bash_profile` (or equivalent) file. It requires no new software, only documentation.

**Comment by dignifiedquire** [comment]
- I think the fix today should be documenting the environment variable and get everyone who cares about this to use it.
- In addition I would like to see `paramfetch` to require a `cache-path` option (or the mentioned env variable) to make sure users of this cli tool do not run into surprises

**Comment by acruikshank** [comment]
@laser 
The only thing I'm really concerned about is that I don't think implementations like go-filecoin should need to provide config for this feature that is otherwise completely hidden. From that standpoint an environment variable makes the most sense to me.

Some alternatives that occurred to me while reading:
1. It seems like the /var directory is a better place for a large cache like this. It would lose the automatic cache eviction, but it's not clear that the OS's /tmp deletion schedule is the correct schedule for deleting groth params.
2. `paramfetch` could create a symlink to the params directory in a consistent location in lieu of other configuration.

I'm not suggesting I've fully thought through these options.

**Comment by laser** [comment]
@acruikshank 

> It seems like the /var directory is a better place for a large cache like this.

That's a good idea. `/var/tmp` is:

1. A directory that exists and is writable on OSX and all Linux flavors I'm aware of
1. Not wiped out between reboots like `/tmp` typically is
1. Not held in RAM, as `/tmp` can be ([on ArchLinux, for instance](https://wiki.archlinux.org/index.php/Tmpfs))

So, I update my proposal: We should modify `paramfetch` and `paramcache` and `parampublish` and `rust-fil-proofs`'s default directory to be `/var/tmp/filecoin-proof-parameters` instead of `/tmp/filecoin-proof-parameters`.

**Comment by anorth** [comment]
@laser I'm sorry I didn't see your message of six days ago until now.

I would have said that I accept your proposal of using an environment variable, on the condition that the environment variable is mandatory: that is, the user *must* actively specify it, and thus make their own decision about /tmp eviction policy, etc.

Using `/var/tmp` is also much better than the current state, though.

This issue title talks about explicit selection of parameters, which I gather is addressed in #721. Is there follow-through work to be done in go-filecoin to set reasonable developer defaults (of no proving params)? If so, please file an issue for us.

**Comment by anorth** [comment]
Also note, when go-filecoin's dep is updated to include this, the documentation at https://github.com/filecoin-project/go-filecoin/wiki/Getting-Started#installing-from-binary should change to reference the right path. I think we should document the environment variable too.

**Comment by porcuquine** [comment]
Quick note: I agree with @schomatis that it makes sense to standardize on the same prefix used for config variables elsewhere in `rust-fil-proofs`. If we can do this when/before documenting, that would make sense — even if we don't yet actually absorb this variable into the config mechanism otherwise used.

**Comment by laser** [comment]
I have created two follow-on issues for this:

- [Change `FILECOIN_PARAMETER_CACHE` environment variable to `FIL_PROOFS_PARAMETER_CACHE` as per convention](https://github.com/filecoin-project/rust-fil-proofs/issues/741)
- [ force consumers of rust-fil-proofs (paramfetch, Rust lib, C lib, paramcache, parampublish) to set Groth parameter cache environment variable](https://github.com/filecoin-project/rust-fil-proofs/issues/742)

CC @porcuquine @dignifiedquire @anorth
